### ISSUE TITLE
Can we double blocksize at the same time as we introduce Segwit?

### ISSUE THREAD

**Issue by GiverofMemory** [dogecoin/dogecoin#1849]
Example code pull https://github.com/bitcoin/bitcoin/pull/6451

This topic has received massive upvotes with 97% up for doubling blocksize when introducing segwit: https://www.reddit.com/r/dogecoindev/comments/mvdzls/transaction_speed_and_fees_hyper_doge/

And this topic describes it also:  https://www.reddit.com/r/dogecoindev/comments/muvl1t/scaling_bitcoin_was_supposed_to_implement_segwit/

Sporkin posts that relate to this: (Thanks u/jamiereeno) 
https://www.reddit.com/r/dogecoin/comments/be8pe8/fully_exited_all_positions_in_doge/el4v0bj/?utm_source=share&utm_medium=ios_app&utm_name=iossmf&context=3

Sporklin post with an acknowledgment from langer_hans: (Thanks u/jamiereeno)
https://www.reddit.com/r/dogecoin/comments/2hdjuj/dogecoin_is_an_independent_codebase/?utm_source=share&utm_medium=ios_app&utm_name=iossmf

**See comments for a proposal by @rnicoll to reduce blocktime to achieve the same goal**

Even supposing Segwit leads to a fully functional lightning network, still a transaction pair needs to be completed before a lightning channel can be opened.  So it makes sense to scale blocksize to accommodate for mass adoption so that fees and congestion stay low - and high value transactions can be settled on-chain.  Also I strongly suggest that segwit addresses always remain optional, like in Bitcoin. My suggestion is to double blocksize whenever needed, perhaps every 7 years.  I think doing this at the same time as the addition of segwit can help set a culture that we will increase blocksize as needed to scale to be money for the world.  Currently we never fill up blocks which also shows that increasing the blocksize isn't going to magically make the blocks bigger in the near term and there would be no difference to running a node.  As a global payments network, we will eventually need to settle (on-chain) hundreds or more transactions per second (even if there are thousands of lightning per second) and we can only accommodate 33 with a 1mb block size.  If we doubled blocksize every 7 years on average, this stays well below moore's law and running a node will never become burdensome to someone willing to invest in a reasonable server.  Again each server is hosting a global payments network so we shouldn't assume a full node will always be able to be run from a smartphone, so I think slowly increasing blocksize would be the way to go to make sure we are able to scale properly.

I understand this will take a hard fork, but it keeps with the core of Dogecoin so it will get full adoption and will not lead to a chain split.  I think the biggest risk to a chain split is refusing to increase blocksize, and that is what led to the creation of Bitcoin Cash.  So a hard fork to double blocksize will not lead to a chain split and can help unify the community and show the world that we indeed will scale to be money for the world and can increase mainstream adoption even more.  ETH hardforks regularly and this does not lead to chain splits, so it is a very viable strategy especially if planned long in advance.

**Comment by shibe2** [comment]
I'm glad if Dogecoin community supports increasing througput of the base network. However, now is not a good time to make that decision. Dogecoin network struggles as it is. We have a lot of work to do to improve it in order to even reach the currently allowed throughput. You cannot increase the capacity by just changing one number. So here is the deal: improvements first, then decisions to change the protocol.

As for SegWit and second-layer solutions, I don't count on them for increasing throughput. Maybe they will help, maybe not. I see it as a separate question.

**Comment by GiverofMemory** [comment]
It is an important question because this is where we choose whether we follow in BTC's footsteps, BCH's footsteps, or do the best of both worlds which is Segwit2x.  If we don't double blocksize at the same time as segwit, it will never happen, just like BTC.

And yes, it is just changing 1 number and would effect literally nothing in the near term, and in the long term would allow TPS to grow.

**Comment by shibe2** [comment]
I don't think so. The thing with BTC is that they simply did not want to increase the throughput of the base network. They used SegWit, Lightning and anything else as excuse for not increasing the block size limit, while in reality these things are not substitutes for each other. We don't have the same problem because Dogecoin community has different mindset.

**Comment by GiverofMemory** [comment]
I don't know if it does.  The community wants the best of both worlds (BTC and BCH) but the dev's here seem pretty set on following in BTC's footsteps...

**Comment by shibe2** [comment]
Dogecoin Core uses Bitcoin Core code base. It doesn't mean that Dogecoin technology moves in the same direction. But it means that, for example, if we wanted SegWit, it's already in the code. If we wanted something that is not there, we'd have to take it from elsewhere or implement it ourselves.

**Comment by GiverofMemory** [comment]
It's a pull request on bitcoin... https://github.com/bitcoin/bitcoin/pull/6451

**Comment by Drael64** [comment]
I certainly understand the want to optimize the existing efficiency first, before expanding it. Approaching a consensus on the issue does not however necessarily mean putting it ahead of that task in priority. 

We could consider it implied that needed optimization tasks fall ahead of block size changes, yes? I've noted that lightening as it exists for BTC today, isn't quite as refined as it could be, whilst segwit does open up multiple future options (such as crosschain, and l2 scaling) and is thus desirable, it doesn't really immediately solve for scaling from what I can determine. 

In my mind lowering fees, and improving TPS are core functionality aims, with dogecoin being conceived by the community to be an alternative cash system. Block size is clearly an obvious choice for TPS. 

Setting up the notion of an incremental increase over time actually isn't a bad idea, as it reduces some of the issues one might have maintaining the network in the short run, and also sets an expectation that helps maintain that keeping of pace of the network with the expansion of block. It doesn't hurt that 7 years is actually dog years ;)

**Comment by rnicoll** [comment]
So immediate thoughts:

* What problem are we solving?
* What does this do to network scalability; bandwidth per node, obviously, but do we need to expand mempool if we're increasing throughput? There's a LOT of people already struggling to download the blockchain (partly P2P doesn't scale aggressively when the network size goes 10x, but partly downloading 100+GB takes a while).
* How does this affect disk usage? I mean the blockchain's already fairly large, are we okay doubling the increase rate? Potentially increasing it exponentially?
* Are there any feasibility issues with miners, especially given they're frequently located in China, and how does the Chinese firewall affect this?
* Does this affect miner viability? If a block is twice the size, it takes twice as long to relay, so is that twice the risk window of two miners creating competing blocks?

Bitcoin had a massive holy war over this, and I don't necessarily agree with their conclusion that basically 10MB every 10 minutes is enough, but we're 10MB every minute. For scaling past that, personally I'd rather we scale horizontally onto multiple chains. We already have pDoge/renDoge wrapped tokens on Ethereum, introducing SegWit means we could look at Lightning (although that has different issues around whether you need to have a license to run a Lightning node, in some places). I accept it's a harder route, but I think multi-chain is a better idea. WDYT?

**Comment by Drael64** [comment]
What problem are we solving?

I'd say transactional speed, from a complete layman's perspective. Dogecoin needs to be as musk eloquently put it 'ideally orders of magnitude faster' to suit it's proposed use as a digital currency. That might not be fully achievable onchain, but there's the immediate rise it use for retail to consider, and the nessasary interaction of crosschain or offchain applications with the mainnet in the longer term (whether that's something like a payment chain, or another chain entirely)

Re: multi-chain. I've thought about this. But there are issues here too. For one, centralized bridges are centralized. 

Secondly they add user friction - you have to manually bridge, you can't send to a dogechain wallet from said bridge, neither of which are intuitive and set up boundaries to adoption. Any crosschain solution needs to interact directly with the native chain in order to be frictionless (something more like thorchain for eg, which is based on cosmos)

Thirdly there's the issue of longevity. Assuming dogecoin is still around in say twenty years, doesn't mean than ANY other chain still will be, or maintained to it's current level. If the chain we rely on for efficiency is in a poor state further down the track, we've introduced a dependency. Things may seem stable in the short term, but the future is major guess work. You could end up building out a whole support ecosystem on another chain, only to find it falls apart. 

I can totally see the reasoning, and desire to do settlement crosschain. Especially because some such chains are orders of magnitude faster already than proof of work (although perhaps less secure, at least theoretically in some cases). And perhaps all these issues can be eliminated in some manner. But I don't think a wrapped ethereum token is close to addressing those issues.

**Comment by Drael64** [comment]
I guess the other thing worth considering is that most of the issues are related to the file size. If the blocks aren't being filled currently, is it actually correct to assume they suddenly will be in technically relevant time frames if the block size is increased? Or is it a non-problem? 

Also I guess, BCH has done this already - what was their experience, and did they have to solve for any of this? That may not be relevant, I'm not sure.

**Comment by GiverofMemory** [comment]
So Gaven Andresen has proposed both compression for BCH as well as Bootstrapping for faster syncing of new nodes.  Both of those could be looked at of course and I seem to think we could get Gavin to help us a bit.

**What problem are we solving?**

The main problem I see us solving by increasing blocksize at the same time as segwit, is setting a culture to both scale on and off chain as reasonable.  I can see devs getting sucked into just following all BTC update path and not really stopping to consider what might or might not be suitable for dogecoin.  If we want to be the peoples money it should always be cheap and fast to settle on-chain, and off-chain also if possible.  We want the best of both worlds.

**What does this do to network scalability; bandwidth per node, obviously, but do we need to expand mempool if we're increasing throughput? There's a LOT of people already struggling to download the blockchain (partly P2P doesn't scale aggressively when the network size goes 10x, but partly downloading 100+GB takes a while).**

Firstly we have to realize that each full node is literally a global payments network.  We run thousands of copies of it compared to Paypal that uses server farms to run ONE copy of it.  It is inevitable that at least a rudimentary server would be required to run a global payments network.  We need to prioritize the end user paying for a stick of gum at circle k more than if it is going to require a $1000 server to run a full node in the future.  It is awesome that anyone on a laptop can run a server currently, but we can't let that hang us up from being a global payments network first and foremost.  Gavin Andresen (successor to Satoshi) has proposed both compression solutions and also Boostrapping solutions to make syncing a new node as painless as possible for BCH.  Also I honestly think that most of our blockchain is a layer-2 to litecoin's merged mining might be playing a large part of why it is tough to sync currently.  As we win more blocks outside of AuxPoW, I think that will improve our syncing rates.

**How does this affect disk usage? I mean the blockchain's already fairly large, are we okay doubling the increase rate? Potentially increasing it exponentially?**

Firstly this will not effect the blockchian size at all in the near term.  This is because we have never filled blocks up entirely.  So even if there was no limit to how large blocks could get, it wouldn't effect anything presently.  As we grow and start to fill up blocks then yes it will grow faster, but only 2x as fast, and it is now 7 years since we started and everyones harddrive space has more than doubled in that time.  I think a very-sub-moore's law scaling (moores law is doubling every 2 years, I am proposing doubling every 7) will keep us safe on disk usage.


**Are there any feasibility issues with miners, especially given they're frequently located in China, and how does the Chinese firewall affect this?**

I have never heard of any problem with this from the BCH people, but would be valuable to reach out and see what SV and BCH devs think.


Does this affect miner viability? If a block is twice the size, it takes twice as long to relay, so is that twice the risk window of two miners creating competing blocks?

Yes there might be a slight increase in orphan rate, which is the measure of this sort of thing.  Orphan rate of dogecoin has reduced dramatically since launch, look at the ANN on bitcointalk.org and when dogecoin started everyone was complaining about orphans!  This relay rate has improved dramatically since and I don't think dogecoin has any issues with orphans now.  I have seen coins with 30 second blocktimes doing ok on orphan rate, so we would be safe.  And no it is not quite a 1:1 time thing, a 2 mb block would not take double as long to relay, maybe a percentage slower like 25-50% I would guess.  There are probably formula's out there.  I will look into our current orphan rate and relay time equations and get back to you.

**Comment by GiverofMemory** [comment]
I can't find doge's orphan rate, but for comparison CranePay has 30 second blocktimes and has around a 1% orphan rate which to me is acceptable. http://cpu-pool.com/tbs  Eth I believe has very short blocktimes and they are doing just fine and their amount of transactions will continue to dwarf ours for a long time.

As you can see here: https://tradeblock.com/blog/bitcoin-network-capacity-analysis-part-6-data-propagation doubling the block size does NOT double to time to relay.  Probably more like 50% longer instead of 100% longer.

**Comment by shibe2** [comment]
>What problem are we solving?

We don't know how fast Dogecoin will be growing. We want to be ready to handle the increased demand if it happens.

>What does this do to network scalability

We need to test that. I'm afraid that even if we removed block size limit completely, we would still not reach the desired throughput because, for example, much of current code runs in a single thread.

>How does this affect disk usage?

Current limit allows about 500 GB of new data per year. That already would prevent most regular users from using Dogecoin Core. One problem with that is that we currently don't have a good open source alternative.

>Are there any feasibility issues with miners, especially given they're frequently located in China, and how does the Chinese firewall affect this?

Miners will probably have to make sure they are directly connected to most other miners. Any intermediate node may have an effect on profitability. They may have VPN tunnels where needed. Making high bandwidth channels to China should be doable.

>Does this affect miner viability? If a block is twice the size, it takes twice as long to relay, so is that twice the risk window of two miners creating competing blocks?

It will not affect viability significantly if propagation times will still be low enough. New blocks normally are not transmitted in full, but are reconstructed on the receiving node. Though we may need to further improve this technology.

>For scaling past that, personally I'd rather we scale horizontally onto multiple chains.

We may eventually having to rely on sharding or off-chain transactions, but we will likely still need high throughput of individual chain. I think that it makes sense to work on increasing on-chain capacity now, which will have immediate benefit in case of increased demand.

**Comment by GiverofMemory** [comment]
Thanks for your input Shibe2!

"Current limit allows about 500 GB of new data per year. That already would prevent most regular users from using Dogecoin Core. One problem with that is that we currently don't have a good open source alternative."

Yes, and again this shows that we are not currently filling up blocks so this change would mean nothing for the short term.  Eventually yes a server size storage amount would be needed, but again I think our priority should be end user experience, not people who want to run a payments network server, which would need to be at least a rudimentary server to serve the globe.  Each node is basically competing with Paypal, so it would be naive to think you could run it on an off the shelf laptop forever.

**Comment by rnicoll** [comment]
Okay, so if we want transactions to be faster, it feels like block time needs to come down, not size per block going up.

Definitely for now I want to even just get 1.21 working (with the tests passing, which is the tricky part, I'm aware there's a prototype that sync but breaks most of the tests, so we have low confidence in it), and get a release cycle rolling, and then worry about what needs to be in the later releases.

**Comment by shibe2** [comment]
Where we need faster transactions, I'd say they should reliably go through in under 5 seconds to compete with other payment options. The idea is to use some kind of off-chain transactions for these cases. For example, based on payment channels.

**Comment by Drael64** [comment]
Yes, 5-10 seconds seems about right for point of sale transactions (like a person, in a store standing in a queue). Under 5 ideally. Not sure if it's practical to lower block time that much. I'd guess not anyway. 

I hope this isn't me showing my stupidity, but I'm going to share some random thoughts. 

1) Say a person locks away 1000 doge in the system, as a bond or stake and then the network confirms their 'fast doge' transaction without actually completing finality (just literally that the transfer is processing), but if it somehow doesn't process or is an attack it uses the stake to compensate the other party. Still using proof of work, but handwaving it for point of sale purposes because the individual has collateralized it?  

Once any pending transactions are fully finalized, the stake money can be withdrawn. Might be a little unwieldy to users. Chunky perhaps. 

2) Would a hybrid consensus method simply for quick payments be an insane or overly complicated idea? I think there are some coins that do that. 

3) There may be a way to solve those xchain issues, such as using wrappers, wallet send/receive incompatibility and dependency on specific chains that I raised. 

I notice thorchain and cosmos use tendermint to achieve some level of native crosschain interaction without wrappers, or manual bridging by the user. Could this be used to create some kind of generic compatibility for a crosschain settlement channel, using other chains for settlement but without being chain specific or requiring that user friction? Leaving the problem somewhat third party, but without the friction or dependency? 

If segwit is still in a state of relative immaturity, then for crosschain purposes it might pay to focus somewhere else anyway. 

These are probably all downright foolish suggestions, but I thought perhaps there might be a way to approach this problem from a fresh angle, rather than using what others have done. I'm not a coder I dropped out of C++ level 1! But maybe there's something in my throwing things at a wall, that inspires a more informed chain of thought.

**Comment by GiverofMemory** [comment]
> Okay, so if we want transactions to be faster, it feels like block time needs to come down, not size per block going up.
> 
> Definitely for now I want to even just get 1.21 working (with the tests passing, which is the tricky part, I'm aware there's a prototype that sync but breaks most of the tests, so we have low confidence in it), and get a release cycle rolling, and then worry about what needs to be in the later releases.

So, its a tradeoff when we are talking blocksize vs blocktime.  Blocksize doubling will not (quite) double to time it takes for the block to propagate the network.  Halving the blocktime will exactly half the time the block gets to propogate before an orphan.  So it seems to me the best gains in terms of scaling for maintaining a low orphan rate is to increase blocksize, not reduce blocktime.  

The benefit to reducing blocktime is faster confirmations, which is a big bonus.  So for raw TPS, blocksize increase is best.  For quicker confirmations and a moderate increase in TPS, lower blocktime is best.  In reality there would be an optimum blocksize and blocktime, no one has figured it out yet.  Another coin with a quick blocktime is Digibyte with 15 second blocktime.

Either way, 30 sec blocktime OR 2mb blocks would turn out to be fine in practice.  the 30 sec blocktime would show results immediately, as people can send transactions faster.  

Another thing to consider is there is a practical limit to blocktime, and you can never really go under 15 seconds because it takes light 0.15 seconds to travel around the world https://web2.0calc.com/questions/how-long-does-it-take-for-light-to-travel-around-the-world , so that is 1% of 15 seconds.  Right now we are very far away from the speed of light connections so we should not go below 30 seconds blocktime IMO.

In conclusion blocksize can (in theory) scale to infinity.  Blocktime cannot.  But we do have room to reduce blocktime, perhaps that is a good step to take since there are much faster coins out there, and Doge was designed to be super fast.

Here is the ANN (which is hard to find) for posterity sake https://bitcointalk.org/index.php?topic=361813.0

**Comment by GiverofMemory** [comment]
> Where we need faster transactions, I'd say they should reliably go through in under 5 seconds to compete with other payment options. The idea is to use some kind of off-chain transactions for these cases. For example, based on payment channels.

That is just not possible, there is a practical limit to how fast blocks can propogate the network.  15 seconds would be the absolute minimum in a p2p currency in my educated opinion since light takes 0.15 to navigate the globe and that is 1% of 15 seconds.  We need there to be a 1% orphan rate (or less) imo.  For a centralized network 5 seconds can work.  Not for decentralized.

**Comment by shalexih** [comment]
I dont see the need for faster block times since zero confirmation
transactions are generally safe. We dont have RBF. Follow what BCH is doing.

**Comment by GiverofMemory** [comment]
```
I dont see the need for faster block times since zero confirmation
transactions are generally safe. We dont have RBF. Follow what BCH is doing.
```
Even so, it is nice to get it in a block, as that does add a good bit more security.  Also halving the blocktime will double TPS just like blocksize doubling would.  Eventually the only way to scale to "infinity" is with blocksize increases on a regular basis.

**Comment by vertiond** [comment]
RBF has been in Dogecoin for over 3 years?

**Comment by Drael64** [comment]
I suppose because dogecoin has only recently waken from it's slumber, there might be network reasons from a developers perspective for favouring block time over size, for now? (ie, some nodes not connecting or propagating well)

And what giverofmemory is advocating (correct me if wrong), is not nessasarily immediate block size increase, but a long term commitment to onchain scaling. 

I can see the argument. Offchain scaling will require onchain scaling to be efficient anyway, as there does need to be interaction with the mainnet, yes? 

And there are vagueries around how well present day solutions, like lightening actually work (ie for btc) and I guess also basically delivery - who will provide them, how decentralized, how secure they are (I've heard people question this with lightening). Perhaps that might tie in also with my question about whether segwit is actually optimal for increasing multichain compatibility, instead of something custom based around tendermint implementation which seems to be the currently popular crosschain methodology? (that might be gobbledegook, but I'm trying!) 

So it seems to me that, in part, it's not about whether increasing block size is a good idea, generally, from a future perspective assuming dogecoin useage grows, just as to whether it's the most practical right _now_. 

Perhaps time could be halved soon-ish (next update or similar), with the intention to increase size later with the growth of the network and the refinement of various levels of optimization? Or that block size increases could be planned, but just deferred until appropriate?

**Comment by shibe2** [comment]
I consider unconfirmed transactions risky, especially considering the current state of Dogecoin P2P network. We could improve reliability of unconfirmed transactions such that they would be acceptable for more use cases, but it would still not be a general solution. We would still need off-chain transactions to improve speed in other use cases.

**Comment by GiverofMemory** [comment]
`And what giverofmemory is advocating (correct me if wrong), is not nessasarily immediate block size increase, but a long term commitment to onchain scaling.`

Yes!

`Offchain scaling will require onchain scaling to be efficient anyway, longer term, as there does need to be interaction with the mainnet, yes? It avoids a certain level of dependency on a relatively unknown solution.`

Yes, and to me I see no reason to implement Segwit without BTC having proven out a lightning network.  Segwit is a bridge, and I prefer not to build a bridge to nowhere without a realized destination.  But if we commit to on-chain scaling, then I am fine with introducing Segwit if that is what people want.  It might have some negative repercussions in the near term though since we are already having issues with syncing and segwit is splitting our network into an entirely other layer which will harm syncing.

`When segwit was written, there were no or next to no active crosschain developments, or actually active networks. We are in a place where we can see these actively popping up and have a better idea of what connections between blockchains look like.`
`If there was a mechanism, within the chain, for commiting a stake of dogecoin, that might allow other networks with tendermint, to apply smart contracts to the locked dogecoin pool?`

Also I'm pretty sure we can use something like OP_RETURN to do cross chain stuff with our current tech, I'm curious what BCH is doing since they have tokens and the like there.  I think the issue was BTC is dead set on keeping 1mb blocks and extra metadata on the blockchain is a big no-no for them hence why they need something like segwit to allow for things outside of blocks.

**Comment by Drael64** [comment]
"Also I'm pretty sure we can use something like OP_RETURN to do cross chain stuff with our current tech, I'm curious what BCH and BSV are doing since they have tokens and the like there. I think the issue was BTC is dead-set on keeping 1mb blocks, and since their blocks are full, extra metadata on the blockchain is a big no-no for them hence why they need something like segwit to allow for things outside of blocks."

Hmm, no I was hinting at something more sophisticated, but that I don't quite have the coding experience to refine or articulate well. 

I'll use a practical example. Thorchain uses tendermint (as most new xchain applications do, if not all), to bridge to other chains. But in order for them to create smart contracts, liquidity pools, loans and the like, they've had to create a liquidity infrastructure that stores the other chains assets. It's a lot of infrastructure. That allows them to layer stuff ontop of native assets, without wrappers or manual bridging. 

Now if there were a way, for external smart contracts, and other chains, to access in a particular, limited fashion, the dogechain directly - like a special type of wallet for example, that could then be locked into a smart contract - then all these features could be built ontop of the doge network, without wrappers, AND without special infrastructure on the other end. 

I'm not sure how this would look. Likely you'd need to go for high confirmation transfers, and limit access of the external party to ensure security, avoid double spends and exploitation. 

But something like this could provide the expandability of a settlement layer, and crosschain composability and communication without a high infrastructure requirement or user end friction. 

Rather than just allowing communication, it would enable functionality to cross chains.

**Comment by Drael64** [comment]
Sorry BTW, to go slightly off topic with the whole multichain thing. I've just been having a kind of lightbulb moment with all that. 

Generally I agree that, onchain scaling is more of an ongoing issue, and would be nice to just see a slow march, in some form. Not knowing the technicals intimately, I have no particular preferences, and defer to others. But it always pays to be future minded. 

It's also possible if what I suggest above was implemented, that faster confirms might help, and at least higher TPS in the long run too (see ethereum to what happens when you start to get not just users, but dapps running on your network).

**Comment by Drael64** [comment]
Oh, and so sorry to add to much off-topic, here, but I learnt Solana uses a blockchain clock (based on SHA256) to help govern their security. Seems like a rather more elegant way to prevent double spend on a xchain, or l2 solution than lightening (just limit the timing of transactions out from those specialized wallets with smart contract access, or otherwise locked funds). 

Adds slightly to my thinking out loud about timeout periods, instead of lightnings fixed payees - that it is possible to keep time on the blockchain.  With a built in chain clock, one could just set the rate at which out transactions can occur via this mechanism to whatever finality is required to prevent double spend.

**Comment by GiverofMemory** [comment]
All good stuff and the concepts you are bringing up we need to talk on discord.  Does Dogecoin have a discord?  That would help.

Solana doesn't have a PoW blockchain.  The clock is probably just to help nodes agree on transaction ordering.  More like a distributed ledger than a blockchain in the sense we think of as a blockchain.  Proof of stake is pretty close to instant.

15 seconds is the absolute minimum for a PoW blocktime.  But we still aren't there yet.  30 seconds is doable with the current network.  For on-chain scaling not only is blocksize increase inevitable to keep a network functioning, it is preferable to reducing blocktime in the sense that for doubling of the blocksize you do not get doubling of transmission time.  So it has better returns on TPS than reducing the blocktime does.  I think both should be done in tandem with small changes over time.

Segwit is fine, it will make node syncing harder.  It won't significantly improve TPS, it is the foundation to a lightning network which will never be more than a niche product.  So to me segwit is more of a distraction from proper TPS on-chain scaling than anything else, and an excuse for devs to tap into that tasty 6.5 million dollar dev fund since segwit would be a major release.

**Comment by Drael64** [comment]
Yeah that's right solana is proof of stake. 

But a blockchain clock could in theory remove the need for confirmations, and also be used as a limiter for offchain, so that offchain transactions coin can be locked only until the onchain is complete? It wouldn't be batch processing, just a timeout. 

So far as I can tell, confirmations preserve immutability of history. But if you have proof of history, via a blockchain clock, perhaps they might not be needed - as the order, or history can be written in to each txn by time based hashes generated by the miners? 

You could also potentially remove orphan blocks this way (earlier completed blocks are preserved, later are removed). Assuming the clock itself is unique and unfalsifiable - ie based around cryptography. It's not simply a timestamp. It's based around generated hashes. 

I might be wildly off on the conceptualization of all this. Maybe it's a dumb/unworkable idea!

**Comment by GiverofMemory** [comment]
Ya we talked about this on discord and it is a good idea.  To preface it, the biggest problem with reducing the blocktime is orphans.  And the biggest problem with orphans is centralization.  
Big miners who mine lots of blocks can complete a mining challenge (1), see that (it looks like) no one else mined it first, and accept their block to mine the next challenge on.  If they win the next block too (2), then their previous block (1) is solidified in the blockchain.  Meanwhile small solo miner actually mined (1) first but by the time it got to big miner, big miner already saw his own block first so rejects the solo miner block (1).  What if there was a way to prove that solo miner actually mined block (1) first?  Well there is and it is called Proof of History (PoH) which Drael64 referenced from Solana.  This (seems to) work by each node broadcasting hashes every 0.4 seconds.  The next broadcast hash needs to hash all the other nodes last hashes.  This way there is this integrated timestamp into everything the miner or node does.  So the miner could include this timestamp hash in his block before mining it, to prove that he mined it when he said he did.  Now that part gets kind of hazy because he could always mine on an old hash and claim that his block was old.  So i'm not quite sure how he can prove something is old.  Not sure how solana solves that.

**Comment by Drael64** [comment]
Yeah, there are some vagueries to how solana actually works to ensure the proofs from the validators (miners in our case), are collated into an absolute sense of time across the chain. I think perhaps that's connected to them being variable (ie can't be predicted in advance). 

But if they do it, it can probably be done for proof of work.

**Comment by raggingsoldier** [comment]
I know I'm commenting a little late but anything that is done that causes unconfirmed transactions on the blockchain needs to be looked over a few times before putting it in effect, if unconfirmed transactions stay in the network for a long period of time it opens the door for malicious attacks on nodes or the transaction itself. Just my two cents worth from work experience

**Comment by GiverofMemory** [comment]
> I know I'm commenting a little late but anything that is done that causes unconfirmed transactions on the blockchain needs to be looked over a few times before putting it in effect, if unconfirmed transactions stay in the network for a long period of time it opens the door for malicious attacks on nodes or the transaction itself. Just my two cents worth from work experience

I don't know what you mean by unconfirmed transactions on the blockchain.  If they are on the blockchain they are, by definition, confirmed.

**Comment by Drael64** [comment]
> I know I'm commenting a little late but anything that is done that causes unconfirmed transactions on the blockchain needs to be looked over a few times before putting it in effect, if unconfirmed transactions stay in the network for a long period of time it opens the door for malicious attacks on nodes or the transaction itself. Just my two cents worth from work experience

If you were speaking about my suggestion, to scale using 'Proof of History' ala solana, but in proof of work (+block time, and eventually block size)- it would be a very large project, and of course there would be testing. 

But the idea is, if time of event/transaction is proofed, in the same way work is currently proofed, then you shouldn't need to do confirmations (or at minimum, far less), there should be less ways to do double spends (or none), and orphan blocks should be fairer. 

Double spend type exploits derive from not knowing which order events come in. Depending on the degree to which you could have a blockchain clock in proof for work (like the tick'), it should at least be able to minimize every one of those. 

Absolutely not sure whether it's entirely workable as a concept either, for proof of work. But seems like as a general concept, it has the potential to solve for a lot of proof of work inefficiencies, all based around not knowing order of events. If order of events could be solved somehow, that reduces the need to scale off chain considerably, long term.

**Comment by Drael64** [comment]
One more thought, and this one more pertains to increasing blocksize. What about a cryptographic accumulator to prove that the current blockchain state is the correct one, rather than a full txn history on nodes? 

This might help mitigate the storage issue.

**Comment by shibe2** [comment]
Reverse download order and UTXO commitment are being worked on in other Bitcoin-based cryptocurrencies. I think that it should be a prerequisite for increasing block size limit.

**Comment by mtrycz** [comment]
As an old time DOGE supporter (we sticked the finger to wolong) and BCH supporter and BCHN contributor, I will post this as a matter of reference for whoever might not have seen it. I hope this can contribute to the decision on scaling.

https://read.cash/@mtrycz/how-my-rpi4-handles-scalenets-256mb-blocks-e356213b

**Comment by SoiledCold5** [comment]
How would increasing the blocksize reduce fees? The feerate is at 1 Doge per KB so increase the blocksize wouldn't help, segwit though would help as it would lower the size of the transaction making the fee less. 

But I am still for a larger blocksize cause it will allow more transactions.

**Comment by shengchalover** [comment]
It’s quite a refreshing experience to read yet another scaling conversation! Conversations are good you all know, so much fun when there is no censorship.

I am mostly a lurker on dogecoin community  and made my bet on BCH some time ago, but I will be totally happy to lose my fortune  to witness an implementation of original Satoshi’s idea as global P2P ecash system. Among thousands coins, dogecoin is one of the very few which has the potential to accomplish this, and I wish the community all the best in finding its path.

Few notes from my experience (I spent quite some time on BTC wallet development in 2017-18 and currently experiment with on-chain social media).

* lightning is truly complicated mess that does not add anything valuable to ecash use-case. It’s nearly impossible to provide a user-friendly experience while maintaining the decentralized nature of the coin. The network does have some interesting applications in e.g. automated (micro)payments for torrent-like networks and similar stuff, but imo we are a decade away from that being useful. As some folks mentioned above, Lightning is better discussed unrelated to scaling.
* block-speed time is very important aspect of adoption from perspective of average users of the network (business and customers). For example, Bitcoin Cash kind of suffers from this (imo) despite huge efforts in making zero-confs reliable. This is mostly a physiological problem but Doge has the potential to overcome it by improving upon its 1min blocks + zero-conf.
* base-layer capacity + block-time are extremely important factors for on-chain social media protocols such as Memo/Member. I believe the transition to blockchain-based social networks is absolutely inevitable and I recommend doge community to lay the foundation for it in a form of bigger OP_RETURN payload + multiple OP_RETURN scripts in a transaction. I work on another implementation of the memo protocol and I would be happy to launch it on dogecoin as well, but 80 bytes is very restrictive (btw, drop me a message on telegram if you are interested in contributing to launching such a network, I need folks to help me in bootstrapping  a webrtc-capable branch of libtorrent + good websocket tracker).

As a final note, my recommendation is to invest in finding the right path to move forward, not just following BTC folks because it’s easier to copy their codebase.

**Comment by shibe2** [comment]
The idea of SegWit is to split transactions in 2 parts. If you count only one part, you have an illusion that the transaction is smaller.

**Comment by SoiledCold5** [comment]
I know, I don’t even like Segwit cause I’m a BCH fan

**Comment by shibe2** [comment]
@RohanKapurDEV Thanks for your insight. Those are important points and I would also like to discuss them elsewhere.

I like the idea of Lightning network, I think that it has important use cases. The only objections I have is against current implementations. If/when some good implementation will come out, I would like Dogecoin to support it.

I like the idea of Segregated Witness, I wish signatures were not shoved into signed data in the first place. The only doubts I have is again about available implementation.

I think that block interval target is already low enough, and further decrease is not worth the trouble. One problem is that actual intervals are random, and sometimes you will be getting long wait time anyway. Another problem is that transactions not always get into the next block. At certain point, transaction confirmation times will not be decreasing significantly, and you will only be getting negative effects. I think that we should aim for reliable confirmations within few seconds, and that's not achievable with block time decrease. So I would bet on some other technology.

Currently Dogecoin has enough capacity for all, but in general, we should aim for success. If another explosion in popularity happens, I want Dogecoin network to be ready for it. Even if explosion will not happen, gradual increase of usage will eventually get us to the point where we will need improvements of the technology. It's always better to prepare ahead of time.

**Comment by Drael64** [comment]
This is another, related idea for scaling. Basically using a proof of stake sidechain for offchain scaling, instead of lightening. Using validation on a side chain, removes the need for payee specific payment channels, which makes payments much more viable. 

https://github.com/dogecoin/dogecoin/issues/1924

**Comment by Landtd2015** [comment]
What problem are we solving?

I see the issue as two fold. The work to be done and the decision to be made.  Block size change, coinciding with Segwit, sets up future increases as a matter of course, assuring the community we are truly trying to be a currency not a wealth storage gimmick.  I can understand the devs just thinking it can be put off because it’s not a priority, but I think the decision would be a signal to the community that our aims remain pure.  We are the peoples’ coin it should be cheap to facilitate txn settlement. 

Also… Multi chain! A layer 2 for retail and institutional entities that want to use doge. Then they could pay the txn fees in the layer two reducing the cross chain traffic and security could be done by the retailer possibly even making a discounted txn fee for running node servers thus removing the onus from the consumer and placing it on the retailer. I know I know we want use case rn but really we are the people’s coin and if we use this as another selling point to the community we will get the use case when they start demanding their favorite retailers accept Doge. You could figure out a way to keep individuals on the main chain thus insuring the value stays in the hands of the consumer. 
Our main focus should be spreading the ethos of the community to the rest of the world if we really want to make this dream come true. I know it’s supposed to be a joke coin but the joke should be that the ethos of the community backing this thing infects the world and we learn to use value for the things that matter not amassing wealth.
You guys ( core devs) are amazing people, you have created a Frankenstein, but this one has friends and a community that believes the stuff you have told us, and if we can truly create a crypto that helps to balance the scales, that’s worth taking the time to think about it, at least, if not working on it this second. I know everyone is super busy rn and we all appreciate it more than you will ever know, and now that the people know you Shibes need help, I imagine you will be getting it in spades soon.

**Comment by ReverseControl** [comment]
I am closing this issue. We now have a discussion space (https://github.com/dogecoin/dogecoin/discussions). I encourage folks who would like to engage on this topic, which is important, to create a discussion to hash out ideas. The Issues section is for tehnical issues we can take action on and not architecture decisions about the blockchain.
